kvbc/src/kvbc_adapter/app_state_adapter.cpp

```diff
@@ -0,0 +1,221 @@
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#include "kvbc_adapter/app_state_adapter.hpp"
+#include "assertUtils.hpp"
+
+namespace concord::kvbc::adapter {
+
+AppStateAdapter::AppStateAdapter(
+    const std::shared_ptr<concord::storage::rocksdb::NativeClient> &native_client,
+    bool link_st_chain,
+    const std::optional<std::map<std::string, categorization::CATEGORY_TYPE>> &category_types,
+    const std::optional<aux::AdapterAuxTypes> &aux_types)
+    : kvbc_(std::make_shared<concord::kvbc::categorization::KeyValueBlockchain>(
+          native_client, link_st_chain, category_types)),
+      blocks_io_workers_pool((bftEngine::ReplicaConfig::instance().numWorkerThreadsForBlockIO > 0)
+                                 ? bftEngine::ReplicaConfig::instance().numWorkerThreadsForBlockIO
+                                 : std::thread::hardware_concurrency()) {
+  if (aux_types.has_value()) {
+    kvbc_->setAggregator(aux_types->aggregator_);
+  }
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// IAppState implementation
+bool AppStateAdapter::hasBlock(BlockId blockId) const { return kvbc_->hasBlock(blockId); }
+bool AppStateAdapter::getBlock(uint64_t blockId,
+                               char *outBlock,
+                               uint32_t outBlockMaxSize,
+                               uint32_t *outBlockActualSize) const {
+  const auto rawBlock = kvbc_->getRawBlock(blockId);
+  if (!rawBlock) {
+    throw NotFoundException{"Raw block not found: " + std::to_string(blockId)};
+  }
+  const auto &ser = categorization::RawBlock::serialize(*rawBlock);
+  if (ser.size() > outBlockMaxSize) {
+    LOG_ERROR(logger_, KVLOG(ser.size(), outBlockMaxSize));
+    throw std::runtime_error("not enough space to copy block!");
+  }
+  *outBlockActualSize = ser.size();
+  LOG_DEBUG(logger_, KVLOG(blockId, *outBlockActualSize));
+  std::memcpy(outBlock, ser.data(), *outBlockActualSize);
+  return true;
+}
+std::future<bool> AppStateAdapter::getBlockAsync(uint64_t blockId,
+                                                 char *outBlock,
+                                                 uint32_t outBlockMaxSize,
+                                                 uint32_t *outBlockActualSize) {
+  static uint64_t callCounter = 0;
+  static constexpr size_t snapshotThresh = 1000;
+
+  auto future = blocks_io_workers_pool.async(
+      [this](uint64_t blockId, char *outBlock, uint32_t outBlockMaxSize, uint32_t *outBlockActualSize) {
+        bool result = false;
+        auto start = std::chrono::steady_clock::now();
+        *outBlockActualSize = 0;
+        LOG_TRACE(logger_, "Job Started: " << KVLOG(blockId));
+
+        // getBlock will throw exception if block doesn't exist
+        try {
+          result = getBlock(blockId, outBlock, outBlockMaxSize, outBlockActualSize);
```

Comment:
I think just this could is implementation-specific and all other code can go into a common wrapper.

Reply:
Thanks @cloudnoize, I have moved the putBlockAsync and getBlockAsync in the common wrapper (replica_adapter)